src/HL7/Message.php

```diff
@@ -365,8 +366,8 @@ public function toString(bool $pretty = false)
         $message = '';
         foreach ($this->segments as $segment) {
             $segmentString = $this->segmentToString($segment);
-            if (!$this->segmentEndingBar) {
-                $segmentString = preg_replace('/\|$/', '', $segmentString);
+            if (! $this->withSegmentEndingFieldSeparator) {
+                $segmentString = substr($segmentString, 0, -1);
```

Comment:
Let's use `preg_replace` instead of `substr` here - it's guaranteed to remove only the field-separator if present and nothing else:
```php
$segmentString = preg_replace('/' . preg_quote($this->fieldSeparator, '/') . '$/', '', $segmentString);
```